docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,263 @@
+
+**C and Objective-C interoperability:** C++ interoperability will have to build off of the previous art of C and Objective-C interoperability, not only to preserve existing Swift codebases and provide a incremental story, but also to reduce unnecessary cognitive burden on developers. Starting with the name, C++ inherits much of C, which means C++ interoperability has a great starting place to build off of. C is a simple language without many high level constructs, so importing C APIs is, relatively, simple. On the other hand, Objective-C adds some high level constructs, such as reference types and a generics model. However, unlike C++, Objective-C has strong idioms; there are only so many patterns for doing some specific thing. So, while C and Objective-C interoperability is a good starting place, the model for importing these languages cannot be taken wholesale for C++ interoperability. 
+
+## Goals
+
+Safety is a top priority for the Swift programming language. C++ often puts safety in the hands of the programmer. Nullability, memory safety, lifetimes, and mutability are all the programmers responsibility in C++. Swift has the opposite default behavior in many of these cases. Simply using C++ APIs should not completely destroy the memory safety, nullability check, etc. that Swift tries to offer. At a minimum, imported C++ APIs should generally not be less safe to use from Swift than they would be in C++, and C++ interoperability should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators). When it is possible for the Swift compiler to statically derive safety properties and API semantics (i.e., how to safely use an API in Swift) from the C++ API interface, C++ interoperability should take advantage of this information. When that is not possible, C++ interoperability should provide annotations to communicate the nessisary information to use these APIs safely in Swift. When APIs cannot be used safely or need careful management, Swift should make that clear to the programmer, and even make those APIs unavailable if there's no reasonable path to a sufficently safe interface.
+
+C++ interoperability should strive to have good diagnostics. When an API cannot be imported, this needs to be communicated to the programmer. The compiler should produce a clear error any time an API that could not be imported is used, and should strive to guide the programmer towards an API interface that can be imported (for example, by suggesting annotations). The compiler should warn users about incorrect API usage, and potentially slow or unsafe patterns. C++ often has complex type names, especially for template specializations, this is an area where the Swift compiler can do better than C++ by showing shorter and more helpful diagnostics, right at the call-site. When the Swift compiler can present better diagnostics than Clang, it absolutely should. 
```

Comment:
```suggestion
C++ interoperability should strive to have good diagnostics. When an API cannot be imported, this needs to be communicated to the programmer. The compiler should produce a clear error any time an API that could not be imported is used, and it should suggest ways that the interface could be changed that would make it importable (for example, by adding annotations). The compiler should also warn users about incorrect API usage, including potentially slow or unsafe patterns.
```

Consider starting with something about it being a goal to provide diagnostics about using C++ APIs that are at least as good as a C++ compiler would provide.  I don't know that you need to elaborate on that with the stuff about ways you could provide better diagnostics — it's interesting, and it'd be good to do, but it's not a core goal, and it's not really in scope for this document.